[Team-13 iOS Neo] 이슈 목록

HoonHaChoi opened the pull request:
안녕하세요 강운! 리뷰 잘부탁드립니다🙇‍♂️

### 작업 목록
 - 이슈 목록 api 모델 구현
 - 이슈 Cell 동적 높이를 가지도록 구현

### 고민과 해결
테이블뷰가 아닌 컬렉션뷰를 이용해서 화면을 만들었습니다.
이슈 리스트의 컨텐츠양의 따라 동적 높이를 어떻게 가지게 할까 고민을 많이 했었는데  더미 Cell을 만들고 해당 컨텐츠에 양에 따른 높이를 받아서  
실제 셀에 높이를 주도록 만들었습니다.

동적 컬렉션뷰를 만드는 도중에 UI 재사용 문제와 메모리 릭 문제를 발견했었는데 
UI 재사용 문제는 마일스톤이나 레이블이 없는 셀을 재사용되는 경우 UI가 올바르게 나오지 않는 현상이 발생하여
 조건문을 통해서 해결 했지만 더 개선하는 방법이 있을지 고민입니다

메모리 릭의 문제의 경우
cell 안에 데이터소스 객체를 생성 했었는데 cell을 재사용 하면서 객체를 계속 생성되는 문제를 발견해  cell 자체에 데이터 소스를 상속하도록 
변경 했고 
레이블 데이터 값을 변경하는 과정에 클로저를 활용해 변경 했었는데 참조 문제 발생시켜 클로저를 제거 하는것으로 문제를 해결했습니다.

궁금한점이 있습니다.
메모리 변화 값, 메모리 그래프를 확인하면서 객체가 쌓이는지 확인하고있습니다.
쌓이는 객체는 없지만 화면스크롤를 반복적으로 위아래로 움직이는 중에 
제가 생각했을때는 메모리가 잘 해제 되었다면 메모리가 다시 원래 값으로 돌아와야 한다 생각하는데 
메모리가 줄다가 기존 값보다 조금씩 메모리가 증가 하는 이유를 잘 모르겠습니다.

HoonHaChoi committed:
feat: URLSession Extension 추가

- 중복을 줄이고 재사용 하기위해 사용

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor:  로그인 Repository 초기화 수정, feat: Issues모델 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: issueListViewModel 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: Model 이름 변경

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: UITabBarController 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: IssueListViewController에 ViewModel 주입

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: 이슈리스트 cell UI 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: CollectionView issue UI(레이블 컬렉션뷰 제외) 동적 높이 구현

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: label 컬렉션뷰 동적 높이 추가, 이슈 동적 높이 수정

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: Custom IssueLabel 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: LeftAlignCollectionFlowLayout 추가

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: IssueListCollectionDataSource 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: LabelCollectionDataSource 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor:  cell identifier extension으로 변경

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: issueCell shadow 추가

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: issues api모델 수정,  milestone 옵셔널 처리 추가

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: Hex String 변환 extension 추가 및 UI 위치 조정 및 label 전달 수정

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: loadingIndicator 추가

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor:  이슈Cell 재사용 문제 수정,  이슈Cell memory leak 문제 수정

- 재사용 문제: 마일스톤,컬렉션뷰가 없는 셀을 재사용하는 문제 발생하여 조건 추가하여 들어온 데이터에 따라 변경되도록 수정
- memory leak 문제: 이슈Cell 스크롤 시 메모리 증가 발생,  문제를 찾는 도중 LabelDataSource를 계속 생성되는 문제 발생 self로 변경, Labels 업데이트 하는 프로퍼티에 약함참조를 추가하여 메모리 릭 문제 해결

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: issueStatusImage 추가

ghojeong/issue-tracker/#101

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/Issue/IssuesListViewModel.swift:
```diff
@@ -0,0 +1,37 @@
+    private var cancellable = Set<AnyCancellable>()
+
+    var errorHandler: ((NetworkError) -> Void)?
+    var issuesList: (([Issue]) -> Void)?
```
클로저에 명사형 이름을 붙이시면 해석할 때 어려움이 있는 거 같습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/TabBarViewController.swift:
```diff
@@ -0,0 +1,50 @@
+        firstTab.tabBarItem = makeTabBarItem(title: "이슈", image: "exclamationmark.circle")
+
+        let secondTab: UIViewController = .init()
+        secondTab.tabBarItem = makeTabBarItem(title: "레이블", image: "tag")
```
enum Item: CaseIterable {
case label ...
}

var items: [String]

init() {
items = Item.allCases.map { $0.description }

"이슈", "레이블"고 같은 데이터는 의미를 부여해주면 코드를 이해할 때 훨씬 더 보기도 좋고 해석할 때 도움을 줄 수도 있습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/TabBarViewController.swift:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    func configureTabBar() {
+        let firstTab = createIssueListViewController()
```
아래 코멘트와 동일하게 스토리보드를 하드 코딩하는 게 아닌 열거형으로 같이 관리했으면 좋겠습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/TabBarViewController.swift:
```diff
@@ -0,0 +1,50 @@
+        let fourthTab: UIViewController = .init()
+        fourthTab.tabBarItem = makeTabBarItem(title: "내정보", image: "")
+
+        self.viewControllers = [firstTab, secondTab, thirdTab, fourthTab]
```
해당 메소드 내부에서 너무나 많은 일이 발생하는 거 같습니다.
의미를 묶을 수 있는 만큼 묶었으면 좋겠습니다.

HoonHaChoi committed:
feat: issueCollectionView refreshControl 추가

ghojeong/issue-tracker/#101

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/Issue/IssueListCollectionDataSource.swift:
```diff
@@ -0,0 +1,38 @@
+        return cell
+    }
+
+    lazy var updateIssues: (([Issue]) -> Void) = { issueList in
```
클로저로 생성하신 이유가 있을까요?

HoonHaChoi replied:
값을 업데이트 하는 간단한 작업을 하는 거라 프로퍼티처럼 만들고 싶었습니다.

이렇게 하게되면 저만 알게 되는 코드가 되는것같아 함수로 변경하는게 좋을거같습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/Issue/IssueListViewController.swift:
```diff
@@ -10,10 +10,11 @@ import UIKit
 
     init?(coder: NSCoder, issueViewModel: IssueListProvider) {
         self.issueViewModel = issueViewModel
+        self.issueDataSourece = IssueListCollectionDataSource()
```
viewModel를 주입시키셨듯이 데이터 소스도 주입하는 게 좋을 거 같습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/Issue/View/IssueCell.swift:
```diff
@@ -48,8 +48,8 @@ class IssueCell: UICollectionViewCell {
-    private var milestoneStackView: UIStackView = {
-        var stack = UIStackView()
+    private lazy var milestoneStackView: UIStackView = {
+        var stack = UIStackView(arrangedSubviews: [self.milestoneImage, self.milestoneLable] )
```
스펠링 틀리셨습니다.

저도 많이하는 실수입니다. 이럴 실수를 줄이는 방법은 커밋을 하기전에 셀프 리뷰를 하는 시간을 갖게되면 해당 실수를 줄일 수 있을 겁니다.

GangWoon commented:
```
동적 컬렉션뷰를 만드는 도중에 UI 재사용 문제와 메모리 릭 문제를 발견했었는데
UI 재사용 문제는 마일스톤이나 레이블이 없는 셀을 재사용되는 경우 UI가 올바르게 나오지 않는 현상이 발생하여
조건문을 통해서 해결 했지만 더 개선하는 방법이 있을지 고민입니다
```

셀이 재사용되는 걸 아셨는데 셀이 재사용되는 시점을 알 수 있는 메소드는 없었을까요?

```
메모리 릭의 문제의 경우
cell 안에 데이터소스 객체를 생성 했었는데 cell을 재사용 하면서 객체를 계속 생성되는 문제를 발견해 cell 자체에 데이터 소스를 상속하도록
변경 했고
레이블 데이터 값을 변경하는 과정에 클로저를 활용해 변경 했었는데 참조 문제 발생시켜 클로저를 제거 하는것으로 문제를 해결했습니다.
```

해당 책임에 대해서 나누는게 좋을 거 같습니다.
이문제도 위 문제와 동일합니다. 이번 고민하신 문제의 키포인트는 셀을 재사용하는 시점에 존재하는 거 같습니다.

```
궁금한점이 있습니다.
메모리 변화 값, 메모리 그래프를 확인하면서 객체가 쌓이는지 확인하고있습니다.
쌓이는 객체는 없지만 화면스크롤를 반복적으로 위아래로 움직이는 중에
제가 생각했을때는 메모리가 잘 해제 되었다면 메모리가 다시 원래 값으로 돌아와야 한다 생각하는데
메모리가 줄다가 기존 값보다 조금씩 메모리가 증가 하는 이유를 잘 모르겠습니다.

```

저도 아직 효율성에 대해서는 아는 부분이 적어서 뭐라고 말씀드리기 어려울 거 같습니다.
셀을 재사용하는 부분에서 생각하신 거 이외에 작업이 발생하고 있는게 아닐까요?

다이나믹 셀 높이를 구현한 코드를 처음 봤는데 얘기는 들어서 알고 있었지만 생각보다 추가적인 작업을 많이 해줘야 하네요...
추가적으로 한가지 부탁을 드리고 싶습니다. 피알의 단위를 체인지 300줄 이하로 제한하고 싶습니다.
매번 많은 양의 리뷰를 진행하기에 정말 부담이 많이가서 그렇습니다.
작업을 작은 단위로 나누실 줄 아는 능력도 면접에서 어필하실 수 있는 중요한 능력입니다.
많은 자료에 대해서 찾아보셨을텐데 고생 많이하셨습니다 :)

HoonHaChoi committed:
refactor: prepareForReuse 메소드 추가

셀이 재사용 될때  cellForRowAt 이전에 호출,  UI 속성을 초기화 하도록 사용됨

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: updateIssues 클로저 에서 함수로 변경

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: milestoneLabel 변수 오타 수정

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: issueListViewController에 issueDataSoure 주입하도록 변경

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: 클로저 핸들러 변수 명 수정

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: TabBarController 열거형 타입에 따라 생성되도록 변경

- TabBarItemType 생성

ghojeong/issue-tracker/#101

HoonHaChoi commented:
많이 코드량 때문에 힘들었을 강운 생각하니 마음이 좋지않고 미안합니다.ㅠ
저도 그러지 않을려고 회사에선 피알 단위를 어떻게 하는지 글들을 찾아보고 배워야겠어요

수정은 했지만 TabBarController 수정한 부분이 이걸 보고 바로 이해가 될 수 있는 코드일까 마음이 걸립니다🤔
그리고 다이나믹 셀 객체의 책임을 나누는 작업은 나중으로 미뤄 놓고싶습니다  현재는 더 고민하고 싶지는 않아요

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/TabBarViewController.swift:
```diff
@@ -9,32 +9,52 @@ import UIKit
-        thirdTab.tabBarItem = makeTabBarItem(title: "마일스톤", image: "signpost.right")
+    private func configureTabBar() {
+        self.viewControllers =
+            tapItems.map { item in makeTabBarController(type: item)}
```
클로저 캡쳐는 항상 조심하셔야해요.
해당 코드 컴파일 되나요?

HoonHaChoi replied:
조심하도록 하겠습니다! 
컴파일 됩니다!

GangWoon commented:
`수정은 했지만 TabBarController 수정한 부분이 이걸 보고 바로 이해가 될 수 있는 코드일까 마음이 걸립니다🤔
`
어떤 이유때문에 그러실까요? 데이터를 관리하실 때 한곳에서 관리를 하시는 게 나중에는 많은 도움이 되실 거에요.
source of truth 에 대해서 찾아보시는 것도 좋을 거 같습니다.

추가적으로 더 이상 고치실 게 없다고 판단되신다면 말씀해주세요~

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/TabBarViewController.swift:
```diff
@@ -9,32 +9,52 @@ import UIKit
-        let thirdTab: UIViewController = .init()
-        thirdTab.tabBarItem = makeTabBarItem(title: "마일스톤", image: "signpost.right")
+    private func configureTabBar() {
+        self.viewControllers =
```
사실 이런 동작들도 ViewModel에서 해주는게 더 자연스럽습니다. 뷰에서 다른 뷰를 만드는 건 뷰와 디펜던시가 밀접한 관계를 갖는 시작점이거든요. 하지만 중요치 않게 생각하신다면 다음 스탭으로 넘어가셔도 괜찮을 거 같습니다

HoonHaChoi committed:
refactor: TabBarController에 각 화면의 코디네이터가 화면이 그려지도록 변경, 불필요한 코드 수정

- UITabbarItem extension 추가
- 각 화면에 대한 코디네이터 생성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: CompositionRoot 생성,  StoryBoarded 프로토콜 추가

- StoryBoarded 스토리보드 init을 조금 더 줄여서 사용하도록 하기위해 작성

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: CompositionRoot 생성, AppCoordinator 생성

- CompositionRoot에서 의존성들을 모두 넣어준다

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: ComposionRoot 로그인과 탭바 코디네이터를 생성하도록 변경
feat: LoginViewCoordinator, TabBarCoordinator 생성 및 프로토콜 생성

- TabBarController 제거

ghojeong/issue-tracker/#101

HoonHaChoi committed:
refactor: AppCoordinaot 변수명 수정, IntialViewController 제거 및 토큰확인 위치 변경

- IntialViewcontroller 할일을 AppCoordinator 하도록 수정
  - AppCoordinator 토큰확인후 화면 이동하도록 변경.
- 화면이 올바르게 보이도록 setViewControllers 변경

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: 로그인완료 Delegate 추가

ghojeong/issue-tracker/#101

HoonHaChoi committed:
feat: 레이블,마일스톤,정보 Coordinator 객체 탭바코디네이터에 추가

ghojeong/issue-tracker/#101

HoonHaChoi commented:
먼저 강운에게 사과부터 하겠습니다.. 부탁하신 300줄 이하는 못 지킨거같습니다.. 미안합니다..
느낌상 수정한 코드가 300줄이 조금 넘지 않을까 싶습니다..

TabBarController에서 다른 뷰 들을 만든다는것이 어색하다는것이 만들면서 많이 느꼈던거같습니다.
그리고  뷰에서 다른 뷰를 만드는 것이 밀접한 관계를 가지게 되어 이것을 고치고자  Coordinator 패턴을 적용 시켜보았습니다

Coordinator 패턴을 적용 시켜보니 뷰가 다른뷰를 알고 있지 않아도 되니 화면 관리를 분리할 수 있는 효과를 봤는데
더 나아가서 
코디네이터객체와 화면 객체, 네트워크 등을 한번에 모아서 관리하기 위해서  Composition Root 에 모든 객체를 한곳에서 생성하도록 만들고 의존성 그래프를 흉내 내봤습니다.

무리하지말고 빠르게 훑터보셔도 됩니다🥲

이미지는 구조 그래프입니다! 
<img width="1275" alt="스크린샷 2021-07-15 오후 3 49 28" src="https://user-images.githubusercontent.com/33626693/125747167-d6c1acec-539e-4121-903f-e406d41c6808.png">

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/Issue/issueListCoordinator.swift:
```diff
@@ -0,0 +1,34 @@
+        self.navigation = navigation
+    }
+
+    func start() {
```
start 라는 이름이 너무 추상적인 거 같습니다.
viewDidLoad, loadInitialData,  등등 다름 이름을 생각해보시는 건 어떠실까요?

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/TabBar/TabBarViewController.swift:
```diff
@@ -0,0 +1,50 @@
+        issueCoordinator.start()
+        labelCoordinator.start()
+        milestoneCoordinator.start()
+        infoCoordinator.start()
```
이런 일련의 작업을 한번에 처리하실 때 속도적인 측면을 신경쓰여야합니다.
예를 들어서 파일 iO, stringliteral 같은 작업은 피하셔야합니다.
지금 코드를 안좋은 습관을 작성하셨다는게 아닌, 위험할 수도 있는 코드여서 말씀드리는 겁니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/Protocol/Coordinator.swift:
```diff
@@ -0,0 +1,20 @@
+ }
+
+ extension Coordinator {
+    func coordinator(to coordinator: Coordinator) {
```
해당 메소드는 무슨의도로 작성하셨나요?

HoonHaChoi replied:
start 함수를 직접 입력하지 않고 
coordinator(to: 해당코디네이터)로 작성하기 위해 썼었는데 사용하지않았던거 같습니다

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/LabelCoordinator.swift:
```diff
@@ -0,0 +1,21 @@
+
+import UIKit
+
+final class LabelCoordinator: Coordinator {
```
코디네이터가 클래스일 이유가 있을까요?

HoonHaChoi replied:
추후에  Delegate 패턴을 이용하기 위해 클래스로 작성했습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/PresentationLayer/InfoCoordinator.swift:
```diff
@@ -9,13 +9,13 @@ import UIKit
 final class InfoCoordinator: Coordinator {
 
-    var navigation: UINavigationController
+    var navigation: UINavigationController?
```
옵셔널일 이유가 있을까요?

HoonHaChoi replied:
조금 더 고민해보고 수정하겠습니다

GangWoon commented:
어떤 부분이 마음에 안드셔서 수정한지 알겠지만, 해당 피알과 무관한 내용을 커밋을 추가하신 거 같습니다. 
지금 상태에서 코드를 확인하려면 제가 커밋을 따라가면서 확인하고, 프로젝트 전체를 다시 확인해야하는 데 너무 비효율적인 거 같습니다..

디테일한 내용은 저녁에 다시 코맨트를 남기도록 하겠습니다.
이번 변경은 다른 피알로 작성하셨으면 좋았을텐데 아쉬워서 했던 말이였습니다.

점차 코드가 점점 책임 분리가 되고 있어서 좋네요. 👍

GangWoon commented on mobile/issue-tracker/issue-tracker/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,62 @@
+import KeychainSwift
+
+protocol AppCoordinatorDependencies {
+    func makeLoginViewCoordinator(navigation: UINavigationController, dependency: LoginViewCoordinatorDependencies) -> LoginViewCoordinator
```
프로토콜 사용하실 때 주의하셔야하는 부분입니다.
추상화를 시켰는데 구차타입을 노출 시키면 추상화시키는 이유가 하나도 없습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,62 @@
+    init(navigation: UINavigationController = UINavigationController(),
+         dependency: dependency) {
+        self.navigation = navigation
+        loginCoordinator = dependency.makeLoginViewCoordinator(navigation: navigation,
```
makeCoordinator(.login) 위처럼 인터페이스를 하나로 통합시킬 수 있을 거 같습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    func start() {
+        if isEmptyToken() {
```
이런 식의 방법은 좋지 않는 방법 같습니다. 컴플리션을 사용하시거나 했으면 좋겠습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/DI/CompositionRoot.swift:
```diff
@@ -0,0 +1,77 @@
+
+extension AppDependency: AppCoordinatorDependencies {
+    func makeLoginViewCoordinator(navigation: UINavigationController, dependency: LoginViewCoordinatorDependencies) -> LoginViewCoordinator {
+        return LoginViewCoordinator(navigation: navigation, dependency: dependency)
```
정말 구체타입을 노출시킬 이유가 있는지 다시 한 버 생각해보셨으면 좋겠습니다.

GangWoon commented on mobile/issue-tracker/issue-tracker/DI/CompositionRoot.swift:
```diff
@@ -0,0 +1,77 @@
+extension AppDependency: TabBarCoordinatorDependencies {
+
+    func makeIssueListCoordinator() -> IssueListCoordinator {
+        return IssueListCoordinator(navigation: NavigationController(),
```
해당 부분도 인터페이스를 통합 시킬 수 있을 거 같습니다.

HoonHaChoi commented:
이번 수정 사항은  짧겠지만  피알을 나누는게 좋겠죠??

GangWoon commented:
네 그러실 수 있을까요? 
머지하고 수정하셨으면 좋겠습니다.
